From: Chris Mason <chris.mason@oracle.com>
Date: Fri, 27 May 2011 10:03:58 -0400
Patch-mainline: yes
References: FATE#306586
Subject: [PATCH] Btrfs: use the device_list_mutex during
 write_dev_supers

write_dev_supers was changed to use RCU to protect the list of
devices, but it was then sleeping while it actually wrote the supers.
This fixes it to just use the mutex, since we really don't any
concurrency in write_dev_supers anyway.

Signed-off-by: Chris Mason <chris.mason@oracle.com>
Signed-off-by: David Sterba <dsterba@suse.cz>
---
 fs/btrfs/disk-io.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.32-SLE11-SP2/fs/btrfs/disk-io.c
===================================================================
--- linux-2.6.32-SLE11-SP2.orig/fs/btrfs/disk-io.c
+++ linux-2.6.32-SLE11-SP2/fs/btrfs/disk-io.c
@@ -2408,7 +2408,7 @@ int write_all_supers(struct btrfs_root *
 	sb = &root->fs_info->super_for_commit;
 	dev_item = &sb->dev_item;
 
-	rcu_read_lock();
+	mutex_lock(&root->fs_info->fs_devices->device_list_mutex);
 	head = &root->fs_info->fs_devices->devices;
 	list_for_each_entry_rcu(dev, head, dev_list) {
 		if (!dev->bdev) {
@@ -2453,7 +2453,7 @@ int write_all_supers(struct btrfs_root *
 		if (ret)
 			total_errors++;
 	}
-	rcu_read_unlock();
+	mutex_unlock(&root->fs_info->fs_devices->device_list_mutex);
 	if (total_errors > max_errors) {
 		printk(KERN_ERR "btrfs: %d errors while writing supers\n",
 		       total_errors);
